### ISSUE TITLE
inline PGP not recognized

### ISSUE THREAD

**Issue by bacuh** [pazz/alot#684]
Just realised that alot is presenting inline PGP as-is. It does no description here.

Is this intentional, a missing feature or just a misconfiguration at my site?

**Comment by sotte** [comment]
:+1: for this feature!

**Comment by elehack** [comment]
ISTM that the tree-of-strings design for message bodies discussed in #639 could be applicable to the presentation problems here — have 'widgets' for signed/encrypted blocks of text, that display a header with formatting that cannot be forged by text.

**Comment by meskio** [comment]
> a) parsing the message is much harder, as one has to get the relevant ciphertext from the message body and feed it to gpg. If pgp/mime is used, the mime container stuff makes this much easier.

Will be nice to parse it and find what you should decrypt. But even if alot doesn't do it it could have a key combination where I can tell it, this is an inline encrypted email try to decrypt it, this is what mutt does.

> b) representing the output is more challenging. imagine an embedded signature and you found and verified it, how do you tell the user about it ? inline is a no-go as this can be trivially forged. presenting the result of the verification in alots header block is also problematic, how do you tell the user which part of the text was signed by which signature ?

Again I think it makes sense the mutt approach for that. They display like `[-- The following data is PGP/MIME signed --]`. Even for PGP/MIME emails it might make sense to display somehow what is signed and what not, there might be attachments that are not signed.

**Comment by laarmen** [comment]
I don't remember how mutt does things exactly, but based on what you said, I could just send you an email pretending to be your boss and tell you to press the nuclear war button, and just put [-- The following data is PGP/MIME signed --] around the line "Push the goddamn button NOW !!".

And, then, KABOOM !

Hence, inline is hard :-)

**Comment by elehack** [comment]
@laarmen Mutt uses color — that cannot be forged by the e-mail sender — to differentiate the signature text from the mail body. Or at least it can, depending on your color theme.

**Comment by josch** [comment]
I have this in the `[bindings]` section of my alot config: `d = pipeto 'gpg --use-agent -d | less'`. This works on all encrypted email I'm receiving here.

**Comment by pazz** [comment]
Excelent. Please document this somewhere more permanent and prominent.
Maybe in the github wiki under "tips" or so.
Thank you!

**Comment by josch** [comment]
And here is the hook that I use to reply to encrypted messages:

    def text_quote(message):
        lines = message.splitlines()
        if lines[0] == '-----BEGIN PGP MESSAGE-----' \
                and lines[-1] == '-----END PGP MESSAGE-----':
            sigs, d = crypto.decrypt_verify(message)
            message = d
        quote_prefix = settings.get('quote_prefix')
        return "\n".join([quote_prefix + line for line in message.splitlines()])

**Comment by josch** [comment]
Thanks, I added both bits to the Contrib-Hooks wiki page. Please modify my edits as necessary.

**Comment by pazz** [comment]
thank you.

**Comment by binaryplease** [comment]
Is this feature planned in near future? 
Not being able to read some encrypted mails properly is a real PITA...

**Comment by pazz** [comment]
Quoting Pablo Ovelleiro Corral (2015-12-01 18:00:53)
>    Is this feature planned in near future?

not unless you're working on it.
Besides, inline pgp is in itself a real PITA :)

**Comment by binaryplease** [comment]
I'm no python pro, but I'd be glad to help if I can and know in which direction to go.

**Comment by zenhack** [comment]
My gut is that as @mdekstrand suggests, the stuff outlined in #639 would be the way to go with this.

**Comment by lamarpavel** [comment]
I sketched out a hook that works for decrypting and reading inline pgp blocks, just like the above mentioned `pipeto 'gpg --use-agent -d | less'`, but the decrypted text is displayed inside alot and without the need for spawning a pager like less.

```
import os

from cStringIO import StringIO
import gpgme


def decrypt_inline(ui):
    # get message content
    msg = ui.current_buffer.get_selected_message()
    msgtext = msg.accumulate_body()

    # decrypt with GPGME
    encrypted_data = StringIO(msgtext)
    plaintext_data = StringIO()
    ctx = gpgme.Context()
    try:
        sigs = ctx.decrypt_verify(encrypted_data, plaintext_data)
    except gpgme.GpgmeError as e:
        raise GPGProblem(e.message, code=e.code)

    plaintext_data.seek(0, os.SEEK_SET)

    # replace message content
    mt = ui.current_buffer.get_selected_messagetree()
    mt.replace_bodytext(plaintext_data.read())
    mt.refresh()

    # refresh the thread buffer
    ui.current_buffer.refresh()
```

My python is shabby and the code is not perfect, but it might serve as template for others to improve. For example `sigs` is sitting around unused while containing valuable information about the verification.

You can make use of this hook by adding a keybinding:
```
[bindings]
	[[thread]]
		D = call hooks.decrypt_inline(ui)
```

Feedback is welcome and might help me to dive in deeper and try making a feature out of it.

**Comment by teythoon** [comment]
Quoting lamarpavel (2016-01-13 21:28:28)
> I sketched out a hook that works for decrypting and reading inline
> pgp blocks, just like the above mentioned `pipeto 'gpg --use-agent
> -d | less'`, but the decrypted text is displayed inside alot and
> without the need for spawning a pager like less.

Yes, but note that the difficulty is not merely to decrypt the text
but to represent it to the user.

**Comment by lamarpavel** [comment]
@teythoon You mean ensuring that the user can positively verify that some text was decrypted? That's why I set up the hook to be called manually by a keybinding so I can trigger it whenever I see an inlined PGP block.

Unless I'm misunderstanding the problem, which might be possible because I just woke up 10 minutes ago.

**Comment by zenhack** [comment]
Having to invoke it manually actually doesn't seem like a terrible idea, and it gives us a solution sooner than if we relied on #639. Other thoughts:

1. This doesn't do the general case, where you may have things *other* than the inline PGP. However, I think in 99% of cases the whole message is one big inline PGP block.
2. If we only support (1), we could actually just render it the same way we do MIME encrypted mail, and do it automatically. This gives an easy answer to the spoofing problem/identifying *what* has been encrypted/signed: everything (or at least the body, which is all you get with MIME anyway).
3. One problem with (2) is it hides whether the message was inline or MIME. I sometimes correspond with folks whose mail clients don't actually support PGP MIME, so I need to be sure to inline messages I exchange with them, which makes that a problem for me.

**Comment by lamarpavel** [comment]
@zenhack, thanks for the reply

To make (1) more robust we could first test whether the entire body is wrapped in the PGP block (sans empty lines), in which case my hook is sufficient and can be applied, while for the other case (inlined PGP blocks _somewhere_ in the body) we don't do anything for now.

For (2), we could simply rename the hook to one of the post/pre hooks and have automatic decryption. But that would lead to another problem: Answering results in a quoted PGP block. The problem here is that you would expect the decrypted text in editor when alot already automatically displays the decrypted text automatically and even verifies the signature.
However, there is a hook in the wiki that seems to solve exactly that specific problem, so I'll have a look at that later.

Which leaves us with (3), to which I have no solution from the top of my head. It's not just a problem for you though, I thinks it's vital to know whether something is PGPMIME or inline out of principle. A half-baked solution would be to manually display the entire header and check the content-type field.

**Comment by zenhack** [comment]
I think having to manually trigger the hook isn't a bad solution to (3). This also makes it harder to forget that we haven't done the stuff related to #639; it would be nice to have that sooner or later.

**Comment by lamarpavel** [comment]
I plan to develop this hook further until it is good enough to be turned into a feature. In that case it should probably work automatically and not be issued manually, right?
Depending on how well this works out,  I'd also start work on the automatic signing/encrypting when replying to signed/encrypted mails. I've just made the jump and comitted to using alot as my main client and working crypto is important to me.

**Comment by lamarpavel** [comment]
The solution for (3) is actually easy, simply add `Content-Type` to `displayed_headers` in your alot config.

Maybe there should be some documented _best practices_ for using alot with crypto where we could assemble this information?

**Comment by teythoon** [comment]
Quoting lamarpavel (2016-01-14 10:23:00)
> @teythoon You mean ensuring that the user can positively verify that
> some text was decrypted? That's why I set up the hook to be called
> manually by a keybinding so I can trigger it whenever I see an
> inlined PGP block.
>
> Unless I'm misunderstanding the problem, which might be possible
> because I just woke up 10 minutes ago.

You are proposing to replace the bodytext in the message tree.  This
is dangerous, because any text in that widget can be trivially forged
by anyone by sending you a mail.

Any text that you display must be entitled using unforgeable visual
elements, e.g. make the pgp markers blue or something.

**Comment by lamarpavel** [comment]
@teythoon That's why my initial idea was to not replace the text automatically but manually. When you receive the mail, you see the PGP block as it is - you then press `D` to invoke `hooks.decrypt_inline(ui)` to replace the text. Either the PGP block is then replaced and you know that it was decrypted, or the decryption fails and the PGP block remains displayed as part of the mail.

**Comment by varac** [comment]
Ugh, 2 years no progress. I'm really missing this feature!

**Comment by dkg** [comment]
I think we should separate this out into "PGP inline encryption" and "PGP inline cleartext signatures"

I think PGP inline cleartext signatures should be completely ignored by every MUA.  they're [just asking for trouble](https://dkg.fifthhorseman.net/notes/inline-pgp-harmful/).

coping with PGP inline encryption can also be risky business, but it's probably less-risky.  the right place to handle this is in libnotmuch itself, though, not in alot.  I've posted [a patch series for dealing with this](https://notmuchmail.org/pipermail/notmuch/2017/025987.html), but it hasn't gotten any significant review.

**Comment by dcbaker** [comment]
I'm also not a fan of inline PGP of any kind. There is no spec of any kind for it, which makes it basically impossible to ensure interoperability, for something so security focused I'm nervous about adding support. Also, its 2018, please stop using non MIME pgp.

**Comment by varac** [comment]
To further split this issue into smaller chunks, I propose to only care for PGP Inline *decryption*.
I also think alot don't need to support pgp inline *encryption*, but I have a hard time convincing ppl that still write me PGP inline mails to use PGP/Mime, and everytime I need to start a new mail client to do so.

**Comment by dcbaker** [comment]
@dkg, that TBE seems really bad. If I understand correctly the problem is that the original encrypted mail is decrypted, then re-encrypted when resent, instead of remaining encrypted at being re-encrypted. Is that correct?

Because we're doing the former with PGP/Mime mail right now.

**Comment by dkg** [comment]
@dcbaker yes, i think your analysis of what *is* happening is correct, though i don't know what you're imagining as the alternative -- what does "remaining encrypted at being re-encrypted" mean (i'm assuming s/at/and/, but even then i still don't get it).

we're talking about *replies*, right?  I don't think it makes any sense to include the encrypted body of the previous message as some sort of sub-part of the current message -- but then again i don't believe that top-posting to people who already got the earlier messages in the thread makes any sense either.

So it's a bit hard for me to get my head around this concern in general, since i believe that a responsible message sender should be aware of all of the parts of the message they are sending.

That said, i think i can imagine the attack you're describing in a PGP/MIME scenario, though i haven't instrumented it myself.  Maybe you could defend against it during reply preparation by the following precautions:

 * only decrypt at most one PGP/MIME-encrypted part as part of preparing the reply
 * if a depth-first walk of the message's mime structure reaches a leaf node before any PGP/MIME-encrypted part is found, do not decrypt any subsequent PGP/MIME-encrypted part.

This is probably something that belongs in `notmuch-reply` though :/

**Comment by dcbaker** [comment]
Okay. Yeah, that seems like a somewhat hard problem to exploit. This also seems pretty much unsolvable, since encrypted something known (in this case the previous mail is a known quantity) tells you something about the encryption used.

**Comment by lucc** [comment]
Is this problem restricted to inline PGP? Can't I prepare some multipart mixed message with two sub parts, the first is text/plain by me (Malory), the second is by Alice end encrypted to bob. Then I hope that Bob will reply to me with Alice message decrypted. 

A more structured description:
1. Alice sends an encrypted (mime) message to Bob
2. Malory copies that message (call it A)
3. Malory crafts a special multipart/mixed message which looks like this:
  * multipart/mixed
    * text/plain: `Hey Bob, please reply, ... \n> Original message from Alice:`
    * encrypted/pgp: A (the encrypted mime part)
4. Hopefully:
  1. Bobs MUA decrypts A and renders all of this as one continuous text
  2. Bob replies and the whole continuous text is pasted in the reply

**Comment by dcbaker** [comment]
No, that won't happen, the way alot does PGP/Mime encryption (correctly) is to always make the pgp-encrypted part be the root part, and any text/* parts will be inside that.

**Comment by dcbaker** [comment]
Oh, and Bob's mail will be encrypted to some number of keys, so only people with keys will be able to read it, everyone else will get an encrypted blob they can't decrypt and no plain text.

**Comment by dkg** [comment]
I think @lucc is right here, that there is likely a similar concern for PGP/MIME.  I think it's marginally worse for Inline PGP, because it's not as straightforward to take the kinds of mitigations i mentioned above.

@dcbaker: the issue isn't how alot does PGP/MIME encryption (which is correct, as you say) -- it's how alot (or any other MUA) deals with maliciously-structured mail when composing its reply.
